keps/sig-apps/20190715-rollingUpdate-enhancement-for-daemonset.md

```diff
@@ -0,0 +1,181 @@
+---
+title: RollingUpdate enhancement for Daemonset
+authors:
+  - "@resouer"
+  - "@zhangxiaoyu-zidif"
+  - "@answer1991"
+owning-sig: sig-apps
+participating-sigs:
+  - sig-apps
+reviewers:
+  - "@janetkuo"
+approvers:
+  - TBD
+editor: TBD
+creation-date: 2019-07-15
+last-updated: 2019-08-05
+status: provisional
+see-also:
+  - n/a
+replaces:
+  - n/a
+superseded-by:
+  - n/a
+---
+
+# RollingUpdate enhancement for Daemonset
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+    - [Story 1](#story-1)
+  - [Implementation Details](#implementation-details)
+    - [API Changes](#api-changes)
+    - [Implementation](#implementation)
+  - [Risks and Mitigations](#risks-and-mitigations)
+  - [Upgrades/Downgrades](#upgradesdowngrades)
+  - [Tests](#tests)
+- [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+- [Drawbacks [optional]](#drawbacks-optional)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Summary
+
+The purpose of this enhancement is to enhance RollingUpdate for Daemonset.
+  - When Daemonset rolling update, it can be `Paused` by user.
+  - Add `SurgingRollingUpdate` for Daemonset as a new update strategy.
+  - Add two more fields, `partition` and `selector` for `RollingUpdate` and `SurgingRollingUpdate`.
+    - `partition` is the number of DaemonSet pods remained to be old version.
+    - `selector` is to query over nodes whoes labels are matched by the Daemonset `RollingUpdate` or `SurgingRollingUpdate`.
+
+
+## Motivation
+
+Consider the following scenarios:-
+
+1. When executing Daemonset update, even those Pods run well, user need to pause updation to check if  some results of new version image meet expect. If not, user can rollback images to stable version.
```

Comment:
How do deployments handle this?

Reply:
sorry, I make it confused. I just want to execute command like using deployment for users.

Reply:
Or, we should provide an observation period(pause for a while) for the upgraded Pod to determine that the upgrade is real and effective. :)

Reply:
it's a good idea to serve an observation channel for users. such as show which node has finished the update and which has not in the whole update process. As to observation period, let me think how to import this feature. ：）

Reply:
@wgliang  we can add some fileds: 
```go
// The total nodes that are running updated daemon pod
    // key is node while value is namespace podName: <Namespace/PodName>.
    // +optional
    UpdatedNumberScheduledNodes map[string]string `json:"updatedNumberScheduledNodes,omitempty" protobuf:"bytes,11,rep,name=updatedNumberScheduledNodes"`
```
to see what the status of nodes and its daemon pod. if you want to pause the sync, you can pasued the action of ds sync.